python/pycrdt/_awareness.py

```diff
@@ -38,164 +36,148 @@ def states(self) -> dict[int, dict[str, Any]]:
 
         Returns:
-            The encoded clients' state.
+            The encoded awareness update.
         """
-        messages = []
+        encoder = Encoder()
+        encoder.write_var_uint(len(client_ids))
         for client_id in client_ids:
-            if client_id not in self._states:
-                continue
-            state = self._states[client_id]
-            meta = self._meta[client_id]
-            update = json.dumps(state, separators=(",", ":")).encode()
-            client_msg = [update]
-            client_msg.insert(0, write_var_uint(len(update)))
-            client_msg.insert(0, write_var_uint(meta.get("clock", 0)))
-            client_msg.insert(0, write_var_uint(client_id))
-            messages.append(b"".join(client_msg))
+            state = self._states.get(client_id)
+            clock = cast(int, self._meta.get(client_id, {}).get("clock"))
+            encoder.write_var_uint(client_id)
+            encoder.write_var_uint(clock)
+            encoder.write_var_string(json.dumps(state, separators=(",", ":")))
+        return encoder.to_bytes()
 
-        if not messages:
-            return None
-
-        messages.insert(0, write_var_uint(len(client_ids)))
-        encoded_messages = b"".join(messages)
+    def apply_awareness_update(self, update: bytes, origin: Any) -> None:
+        """
+        Applies the binary update and notifies subscribers with changes.
 
-        message = [
-            write_var_uint(YMessageType.AWARENESS),
-            write_var_uint(len(encoded_messages)),
-            encoded_messages,
-        ]
-        return b"".join(message)
+        Args:
+            update: The binary update.
+            origin: The origin of the update.
+        """
+        decoder = Decoder(update)
+        timestamp = int(time.time() * 1000)
+        added = []
+        updated = []
+        filtered_updated = []
+        removed = []
+        length = decoder.read_var_uint()
+        for _ in range(length):
+            client_id = decoder.read_var_uint()
+            clock = decoder.read_var_uint()
+            state_str = decoder.read_var_string()
+            state = None if not state_str else json.loads(state_str)
+            client_meta = self._meta.get(client_id)
+            prev_state = self._states.get(client_id)
+            curr_clock = 0 if client_meta is None else client_meta["clock"]
+            if curr_clock < clock or (
+                curr_clock == clock and state is None and client_id in self._states
+            ):
+                if state is None:
+                    # Never let a remote client remove this local state.
+                    if client_id == self.client_id and self.get_local_state() is not None:
+                        # Remote client removed the local state. Do not remove state.
+                        # Broadcast a message indicating that this client still exists by increasing
+                        # the clock.
+                        clock += 1
```

Comment:
Shouldn't it be 
```python 
clock = curr_clock + 1
```
to update the local clock ?

Reply:
This mimics the [JavaScript implementation](https://github.com/yjs/y-protocols/blob/255b7e6c96d9c6dd48b5519b3f83ab65afdb4bba/awareness.js#L262). Should it be different?

Reply:
Maybe I misunderstood the clock, but I understood that each client has its own.
If this is the case, why should we rely on a value coming from a client to update the local clock ?

Reply:
Do you want to open an issue in https://github.com/yjs/y-protocols to discuss this, or maybe @dmonad can comment here?